File: crates/fresh-editor/plugins/env-manager.ts
/// <reference path="./lib/fresh.d.ts" />

/**
 * Environment Manager
 *
 * Detects a project's environment manager (Python venv, direnv, mise) and
 * activates it by handing core an activation **snippet** via `editor.setEnv`.
 * Core captures the resulting environment on the active backend (local / SSH)
 * and applies it to every editor-spawned process — language servers,
 * formatters, `spawnProcess`.
 *
 * Detection is passive (reads files only). Activation runs repo-controlled
 * code, so it is gated on Workspace Trust: the plugin only calls `setEnv` when
 * `editor.workspaceTrustLevel() === "trusted"` (and core enforces the same).
 *
 * ## Activation strategy (see `docs/internal/trust-env-devcontainer-ux-plan.md`)
 *
 * Detected envs are split by what their activation actually *does*:
 *
 * - **path-only** (`.venv` / `venv`): activation is a `PATH` prepend and a few
 *   env-var sets. No arbitrary shell is run against repo-controlled scripts
 *   *that the user authored* — yes, `activate` is sourced, but it's a fixed
 *   script that pyvenv/virtualenv writes. We treat this as low-risk and
 *   auto-activate on plugin load with no popup, mirroring VS Code Python.
 *   Undo is one click on the status pill (or `Env: Use System`).
 * - **shell** (`.envrc` / `mise.toml` / `.mise.toml` / `.tool-versions`):
 *   activation runs `direnv export` / `mise env`, which evaluate user shell
 *   inside the repo. This is the dangerous case, so it is gated on Workspace
 *   Trust. The plugin does *not* ask the trust question itself — the core
 *   workspace-trust modal is the single prompt for that, and it already names
 *   the specific `.envrc`/`mise` marker. When the user trusts the folder, core
 *   fires the `trust_changed` hook and this plugin activates the env in
 *   response. One decision, one prompt, no duplicate "Trust & activate" popup.
 *
 * Coordination with the devcontainer plugin: if a `devcontainer.json` is
 * present and the current authority is local, env-manager defers entirely —
 * the devcontainer plugin's "Reopen in Container?" popup goes first. After
 * the user attaches and the editor restarts under the container authority,
 * env-manager re-runs and asks about the env from inside the container, which
 * is the right place to do it. If the user dismisses the devcontainer popup,
 * env-manager picks up its own decision on the next plugin reload (next
 * editor restart).
 *
 * Freshness: one-shot spawns re-capture automatically when the env inputs
 * change (core's cache is keyed on them). A long-running language server has
 * its env fixed at spawn, so to pick up a changed `.envrc`/`mise.toml` the
 * user runs **Env: Reload**, which re-captures and restarts servers. (Auto
 * file-watching is intentionally not wired yet.)
 */

const editor = getEditor();

const STATUS_TOKEN = "env";

/// Devcontainer plugin's attach-popup id. We listen for its outcome on the
/// shared `action_popup_result` channel so we can un-defer the env popup
/// when the user declines the devcontainer attach — see the
/// `onDevcontainerAttachResult` handler below.
const DEVCONTAINER_ATTACH_POPUP_ID = "devcontainer-attach";

interface ActionPopupResultData {
  popup_id: string;
  action_id: string;
}

interface Detected {
  /** Short label for the status pill, e.g. ".venv" / "direnv" / "mise". */
  name: string;
  /** The activation snippet handed to `editor.setEnv`. */
  snippet: string;
  /**
   * "path-only" envs (`.venv`/`venv`) auto-activate silently when trusted.
   * "shell" envs (`.envrc`/`mise.toml`/`.tool-versions`) only activate once
   * the workspace is trusted (the core trust modal owns that prompt).
   */
  kind: "path-only" | "shell";
}

function fileExists(p: string): boolean {
  try {
    return editor.fileExists(p);
  } catch (_e) {
    return false;
  }
}

/**
 * Detect the environment in the current workspace and return its activation
 * snippet, or null if none. These are auto-detected default snippets; direnv
 * and mise need their exporters (they're prompt-hook driven), venv sources its
 * activate script, and anything else is a pure login shell / user snippet.
 */
function detect(): Detected | null {
  const cwd = editor.getCwd();
  if (!cwd) return null;

  for (const name of [".venv", "venv"]) {
    const dir = editor.pathJoin(cwd, name);
    if (
      fileExists(editor.pathJoin(dir, "bin", "python")) ||
      fileExists(editor.pathJoin(dir, "bin", "python3")) ||
      fileExists(editor.pathJoin(dir, "Scripts", "python.exe"))
    ) {
      return {
        name,
        snippet: `source ${editor.pathJoin(dir, "bin", "activate")}`,
        kind: "path-only",
      };
    }
  }

  if (fileExists(editor.pathJoin(cwd, ".envrc"))) {
    return {
      name: "direnv",
      snippet: `eval "$(direnv export bash)"`,
      kind: "shell",
    };
  }

  for (const name of ["mise.toml", ".mise.toml", ".tool-versions"]) {
    if (fileExists(editor.pathJoin(cwd, name))) {
      return {
        name: "mise",
        snippet: `eval "$(mise env -s bash)"`,
        kind: "shell",
      };
    }
  }

  return null;
}

function isTrusted(): boolean {
  return editor.workspaceTrustLevel() === "trusted";
}

/**
 * Whether a devcontainer config exists at the workspace root. Used to decide
 * whether to defer to the devcontainer plugin's attach popup. We do a passive
 * file check rather than reach across plugins so the two stay independent.
 */
function devcontainerConfigPresent(): boolean {
  const cwd = editor.getCwd();
  if (!cwd) return false;
  return (
    fileExists(editor.pathJoin(cwd, ".devcontainer", "devcontainer.json")) ||
    fileExists(editor.pathJoin(cwd, ".devcontainer.json"))
  );
}

/**
 * True when an authority other than "local" is installed — i.e. the editor
 * is already attached to a container or SSH host. Used by the defer-to-
 * devcontainer rule: we only stand aside *before* attach. After attach, the
 * authority is non-empty and we're free to surface our own popup.
 */
function authorityIsNonLocal(): boolean {
  return editor.getAuthorityLabel().length > 0;
}

// === Cross-plugin: devcontainer decline observation ===
//
// `maybeAutoActivate` defers when a `devcontainer.json` is present on the
// host so the devcontainer "Reopen in container?" popup goes first. The
// risk this introduces — and that this section closes — is that the
// devcontainer popup may *not actually appear* (the user previously
// chose "Ignore always", or already attached and detached, or some
// other skip-path in `devcontainer.ts:2729-2776`). Without a signal that
// devcontainer is *not* going to prompt, the env popup would never
// appear on the host, even though the user is staying local.
//
// Two pieces of state record what we've observed about devcontainer:
//
// - `devcontainerDismissedThisSession` (in-memory): set when the user
//   picks any non-attach option in the devcontainer-attach popup this
//   session. Re-running `maybeAutoActivate` after this flag is set
//   bypasses the defer guard and lets the env popup surface
//   immediately, in the same session.
// - `devcontainer-decline:<cwd>` in plugin global state (persisted):
//   set when the user picks "Ignore always" in the devcontainer popup.
//   On the *next* open of the same folder, devcontainer reads its own
//   persisted dismissal and silently skips the popup; without this
//   observation, env-manager would defer to a popup that never comes.
//   With it, env-manager proceeds straight to its own activate flow.
//
// We can't read devcontainer's own `attach:<cwd>` global state because
// plugin global state is namespaced per plugin (see fresh.d.ts:2700-2710).
// So we keep our own copy of the relevant observation, written when we
// see the user's choice on the shared `action_popup_result` channel.
let devcontainerDismissedThisSession = false;

function devcontainerObservationKey(): string {
  return "devcontainer-decline:" + editor.getCwd();
}

function readDevcontainerDeclined(): boolean {
  return editor.getGlobalState(devcontainerObservationKey()) === "user_dismissed";
}

function writeDevcontainerDeclined(): void {
  editor.setGlobalState(devcontainerObservationKey(), "user_dismissed");
}

// === Commands ===

/**
 * Apply `setEnv` and surface the activating/reloading status message.
 * Pre-condition: trust must already be Trusted (the caller is responsible).
 * Core captures the snippet on the active backend and restarts so language
 * servers re-spawn under the fresh env.
 */
function applyActivation(det: Detected): void {
  editor.setEnv(det.snippet, editor.getCwd());
  editor.setStatus(
    editor.t(editor.envActive() ? "status.reloading" : "status.activating", { name: det.name }),
  );
}

/** Activate (or, when already active, reload) the detected environment.
 *
 * Trust handling: if the workspace is not trusted, we open the *core*
 * workspace-trust modal (the single trust prompt) rather than a second,
 * plugin-owned trust popup. Once the user trusts the folder, core fires
 * `trust_changed` and our subscription below re-runs the activation — so the
 * user's one "Trust" decision both elevates trust and activates the env,
 * without us asking the trust question ourselves.
 */
function activate(): void {
  const det = detect();
  if (!det) {
    editor.setStatus(editor.t("status.no_env_detected"));
    return;
  }
  if (!isTrusted()) {
    requestCoreTrustPrompt();
    return;
  }
  applyActivation(det);
}
registerHandler("env_activate_handler", activate);

function useSystem(): void {
  editor.clearEnv();
  editor.setStatus(editor.t("status.deactivated"));
}
registerHandler("env_use_system_handler", useSystem);

function showStatus(): void {
  const det = detect();
  const trust = editor.workspaceTrustLevel() || "unavailable";
  if (editor.envActive()) {
    editor.setStatus(
      det
        ? editor.t("status.env_active_named", { name: det.name })
        : editor.t("status.env_active"),
    );
  } else if (det) {
    editor.setStatus(editor.t("status.env_detected", { name: det.name, trust }));
  } else {
    editor.setStatus(editor.t("status.no_env", { trust }));
  }
}
registerHandler("env_status_handler", showStatus);

editor.registerCommand("%cmd.activate", "%cmd.activate_desc", "env_activate_handler");
editor.registerCommand("%cmd.reload", "%cmd.reload_desc", "env_activate_handler");
editor.registerCommand("%cmd.use_system", "%cmd.use_system_desc", "env_use_system_handler");
editor.registerCommand("%cmd.status", "%cmd.status_desc", "env_status_handler");

// === Trust prompt ===

/**
 * Open the *core* workspace-trust modal — the single, editor-owned trust
 * prompt. Plugins can't open it through a dedicated API, but the editor
 * exposes `workspace_trust_prompt` as an action and `executeActions` is the
 * generic dispatch channel. We deliberately do NOT show a second, plugin-owned
 * "Trust & activate" popup: that duplicated the trust question in a competing
 * UI. Activation happens as a *consequence* of trust, via the `trust_changed`
 * subscription below.
 */
function requestCoreTrustPrompt(): void {
  editor.executeActions([{ action: "workspace_trust_prompt", count: 1 }]);
}

/// Catch the devcontainer attach popup's outcome on the shared
/// `action_popup_result` channel. Any non-attach action means the user is
/// staying on the host, so we should un-defer and let the env popup
/// surface in the same session. `dismiss_always` is also persisted so
/// the next open of this folder doesn't re-defer (devcontainer will
/// silently skip its popup that time).
function onDevcontainerAttachResult(data: ActionPopupResultData): void {
  if (data.action_id === "attach") {
    // editor.setAuthority restarts the editor; env-manager re-runs
    // inside the container via the post-restart `plugins_loaded`.
    return;
  }
  devcontainerDismissedThisSession = true;
  if (data.action_id === "dismiss_always") {
    writeDevcontainerDeclined();
  }
  // Re-evaluate now that the defer barrier is gone.
  maybeAutoActivate();
}

editor.on("action_popup_result", (data) => {
  // env-manager no longer owns a trust popup; the only action popup it still
  // coordinates with is the devcontainer plugin's attach prompt.
  if (data.popup_id === DEVCONTAINER_ATTACH_POPUP_ID) {
    onDevcontainerAttachResult(data);
  }
});

// Trust is granted elsewhere — the core trust modal, the status-bar trust
// pill, or a palette command — all of which route through core's
// `set_workspace_trust_level` and fire `trust_changed`. That is our cue to
// activate a now-trusted env, so the user's single "Trust" decision both
// elevates trust and turns the env on, with no second prompt from us.
editor.on("trust_changed", (data) => {
  if (data.level === "trusted" && !editor.envActive()) {
    maybeAutoActivate();
  }
});

// === Plugin-load orchestration ===

/**
 * Decide what (if anything) to do on plugin load for the detected env.
 *
 * Routing:
 * - No env detected → nothing.
 * - Path-only (`.venv`/`venv`) → auto-activate silently if trusted. Path-only
 *   is intentionally exempt from trust prompting; the snippet is just `PATH`
 *   setup.
 * - Shell env, devcontainer present, local authority, no observed
 *   decline → defer entirely. The devcontainer attach popup goes first; we
 *   re-run after the post-attach restart inside the container, or when the
 *   user declines the devcontainer popup (see `onDevcontainerAttachResult`).
 * - Shell env, devcontainer present but user already declined the attach
 *   (this session or persistently) → fall through to the env flow on the
 *   host.
 * - Shell env, already activated → nothing (the env is live; user can reload).
 * - Shell env, trusted → silent activation (trust is the green light; honor it).
 * - Shell env, not yet trusted → nothing. The core workspace-trust modal owns
 *   the ask; when the user trusts the folder, `trust_changed` re-runs this and
 *   we activate. We never surface our own trust popup here.
 */
function maybeAutoActivate(): void {
  const det = detect();
  if (!det) return;

  if (det.kind === "path-only") {
    if (isTrusted() && !editor.envActive()) {
      applyActivation(det);
    }
    return;
  }

  // det.kind === "shell"
  if (editor.envActive()) return;
  if (devcontainerConfigPresent() && !authorityIsNonLocal()) {
    // Only defer while the user might still see the devcontainer prompt.
    // If they declined it earlier in this session, or persistently
    // declined in a previous session (so devcontainer is silently
    // skipping its popup), proceed to the env activate flow instead of
    // waiting for a popup that will never appear.
    if (devcontainerDismissedThisSession) {
      editor.debug(
        "env-manager: devcontainer dismissed this session — proceeding with env activate",
      );
    } else if (readDevcontainerDeclined()) {
      editor.debug(
        "env-manager: user previously declined devcontainer attach for this folder — proceeding with env activate",
      );
    } else {
      editor.debug(
        "env-manager: deferring to devcontainer plugin (config present, local authority)",
      );
      return;
    }
  }

  if (isTrusted()) {
    // Trust is granted (now or previously) → activate. Trust is the only gate.
    applyActivation(det);
    return;
  }
  // Not trusted yet: stay silent. The core trust modal asks; `trust_changed`
  // brings us back here once the user trusts.
}

registerHandler("env_maybe_auto_activate", maybeAutoActivate);
editor.on("plugins_loaded", "env_maybe_auto_activate");

// A session activated *after* boot — most notably one spawned through the
// Orchestrator ("New Session" / dock), which creates a window without
// re-firing `plugins_loaded` — must get the same auto-activation a direct
// `fresh <dir>` launch gets. Without this, such a session stays on the system
// toolchain even once its workspace trust is decided (issue #2355 follow-up).
// Guarded on `!envActive()` so merely switching back to an already-active
// session is a no-op and never re-applies (which would needlessly reload LSP).
registerHandler("env_auto_activate_on_session", () => {
  if (!editor.envActive()) maybeAutoActivate();
});
editor.on("active_window_changed", "env_auto_activate_on_session");

// === Status pill (opt-in to a user's status-bar layout) ===
//
// One pill: "env" — what environment is active (always relevant once
// env-manager runs). Workspace trust has its own first-class, always-visible
// core status-bar element (`{trust}`, rendered from the active session's trust
// level every frame and capitalized); env-manager no longer registers a
// competing per-buffer trust chip — that one rendered lowercase and vanished
// for sessions whose buffer hadn't been refreshed.

function refreshStatus(): void {
  const bufferId = editor.getActiveBufferId();
  if (bufferId === 0) return;
  const det = detect();
  let value: string;
  if (editor.envActive()) {
    value = det
      ? editor.t("statusbar.active", { name: det.name })
      : editor.t("statusbar.active_unknown");
  } else if (det) {
    value = isTrusted()
      ? det.name
      : editor.t("statusbar.locked", { name: det.name });
  } else {
    value = editor.t("statusbar.system");
  }
  editor.setStatusBarValue(bufferId, STATUS_TOKEN, value);
}

editor.registerStatusBarElement(STATUS_TOKEN, editor.t("statusbar.label"));

registerHandler("env_refresh_status", refreshStatus);
// `ready` populates the pills at boot (after workspace restore opened the
// initial buffer) so an opted-in trust chip shows immediately, not only after
// the first buffer switch / file open.
for (const event of ["ready", "buffer_activated", "after_file_open", "focus_gained"]) {
  editor.on(event, "env_refresh_status");
}

// === Clickable chip ===
//
// The `env` pill env-manager registers becomes a first-class affordance back
// to its decision. Clicking it re-runs the activate flow, re-opening the
// activate prompt if a pending decision remains — the "Status beats prompts"
// callback, where the indicator is the affordance, not just a passive label.
// (The workspace-trust `{trust}` element is core and owns its own click →
// trust prompt; env-manager no longer handles trust clicks.)
editor.on("status_bar_token_clicked", (data) => {
  if (data.plugin_name !== "env-manager") return;
  if (data.token_name === STATUS_TOKEN) {
    // Click on the env pill is an explicit "act on this env" gesture. If the
    // folder is trusted, activate (or reload). If it isn't, open the core
    // trust modal — the single trust prompt — rather than a plugin-owned
    // popup; trusting then activates via `trust_changed`. If there's no env
    // detected at all, fall back to a status message so the click isn't silent.
    const det = detect();
    if (!det) {
      editor.setStatus(editor.t("status.no_env_detected"));
      return;
    }
    if (isTrusted()) {
      applyActivation(det);
    } else {
      requestCoreTrustPrompt();
    }
  }
});

refreshStatus();

File: crates/fresh-editor/tests/e2e/plugins/env_manager.rs
//! E2E tests for the trust + env open-time UX (see
//! `docs/internal/trust-env-devcontainer-ux-plan.md`). Each test drives
//! keyboard input and asserts only on rendered screen content — never on
//! internal state — per CONTRIBUTING.md's "observe, not inspect" rule.
//!
//! These cover four flows visible to the user when opening a folder:
//!
//! - **Path-only env (`.venv`)**: auto-activate silently, status pill shows
//!   `.venv ✓`, no popup, no trust modal.
//! - **Shell env (`.envrc`)**: the *core* trust modal (the same single prompt
//!   manifests get), with concrete framing naming `.envrc`. Picking "Trust"
//!   elevates trust and env-manager activates direnv in response — driven by
//!   the `trust_changed` hook, not a separate plugin "Trust & activate" popup.
//! - **Project manifest (`Cargo.toml`)**: trust modal fires with concrete
//!   framing that names the actual marker.
//! - **Cancel the trust modal** (T19 — Ctrl+Q quit without picking): the
//!   on-disk trust file must NOT be written, so the next open of the same
//!   folder re-fires the prompt.

use crate::common::harness::{copy_plugin, copy_plugin_lib, EditorTestHarness};
use crossterm::event::{KeyCode, KeyModifiers};
use fresh::config::Config;
use fresh::config_io::DirectoryContext;
use std::fs;
use std::path::{Path, PathBuf};
use tempfile::TempDir;

/// Copy env-manager + its lib into the project's `plugins/` directory so the
/// plugin loads when the harness boots the editor against `working_dir`.
fn setup_env_manager(working_dir: &PathBuf) {
    let plugins_dir = working_dir.join("plugins");
    fs::create_dir_all(&plugins_dir).expect("create plugins dir");
    copy_plugin(&plugins_dir, "env-manager");
    copy_plugin_lib(&plugins_dir);
}

/// Synthetic `.venv/bin/python` so env-manager's `detect()` recognizes the
/// folder as a Python venv (it walks the same file checks live).
fn make_venv(root: &Path) {
    let venv = root.join(".venv").join("bin");
    fs::create_dir_all(&venv).expect("create venv");
    fs::write(
        venv.join("activate"),
        b"#!/bin/sh\nexport VIRTUAL_ENV=/tmp\n",
    )
    .expect("write activate");
    fs::write(venv.join("python"), b"").expect("write python");
}

/// `.envrc` content doesn't matter for detection (env-manager only checks
/// existence); we put something sane so an accidental shell eval doesn't
/// fail loudly during debugging.
fn make_envrc(root: &Path) {
    fs::write(root.join(".envrc"), b"export FOO=bar\n").expect("write .envrc");
}

fn make_cargo_toml(root: &Path) {
    fs::write(
        root.join("Cargo.toml"),
        b"[package]\nname = \"demo\"\nversion = \"0.0.1\"\nedition = \"2021\"\n",
    )
    .expect("write Cargo.toml");
}

/// PTY availability guard — `create_window_with_terminal` spawns a real shell,
/// which the CI sandbox occasionally can't allocate.
fn pty_available() -> bool {
    portable_pty::native_pty_system()
        .openpty(portable_pty::PtySize {
            rows: 1,
            cols: 1,
            pixel_width: 0,
            pixel_height: 0,
        })
        .is_ok()
}

/// Boot the editor harness exactly like a real `fresh /path` launch does
/// at startup. `EditorTestHarness::with_config_and_working_dir` skips two
/// hooks `main.rs` runs immediately after editor construction:
///
/// 1. `editor.maybe_prompt_workspace_trust()` (`main.rs:3902`) — resolves
///    the per-folder trust level. Without this, trust stays at the default
///    `Restricted`, so env-manager's `maybeAutoActivate` short-circuits on
///    the `isTrusted()` check for path-only envs and the open-time trust
///    modal never fires for project manifests.
/// 2. `editor.fire_plugins_loaded_hook()` (`main.rs:3840`) — dispatches
///    the `plugins_loaded` lifecycle event. Without this, env-manager's
///    `maybeAutoActivate` (registered on that hook) never runs at all.
///
/// Wire both here so the test surface mirrors what a user actually sees on
/// `fresh /path`.
fn boot_harness_like_main(width: u16, height: u16, project: PathBuf) -> EditorTestHarness {
    boot_with_dir_context(width, height, project, None)
}

/// Same as `boot_harness_like_main` but takes an explicit `DirectoryContext`
/// so two consecutive boots can share their data dir. Required by the
/// "cancel doesn't leak a decision" test (T19) — without a shared
/// dir_context, each harness gets its own temp `data_dir` and any
/// persisted `trust.json` from boot 1 is invisible to boot 2, hiding the
/// bug the test exists to catch.
fn boot_with_dir_context(
    width: u16,
    height: u16,
    project: PathBuf,
    dir_context: Option<DirectoryContext>,
) -> EditorTestHarness {
    let mut harness = match dir_context {
        Some(dc) => EditorTestHarness::with_shared_dir_context(
            width,
            height,
            Config::default(),
            project,
            dc,
        )
        .unwrap(),
        None => EditorTestHarness::with_config_and_working_dir(
            width,
            height,
            Config::default(),
            project,
        )
        .unwrap(),
    };
    // Wire a per-project trust store so `set_level` calls inside
    // `maybe_prompt_workspace_trust` actually persist to disk. The
    // harness builds `WorkspaceTrust::permissive()` without a store;
    // without this step, T19's persistence assertion can't tell the
    // difference between "decision recorded" and "no store wired" — both
    // look identical to the gate.
    let store_path = {
        let editor = harness.editor();
        let working_dir = editor.working_dir().to_path_buf();
        editor.dir_context().project_state_dir(&working_dir)
    };
    let store = fresh::services::workspace_trust::TrustStore::for_project_dir(&store_path);
    harness
        .editor()
        .authority()
        .workspace_trust
        .set_store(Some(store));
    harness.editor_mut().maybe_prompt_workspace_trust();
    // Republish the plugin state snapshot so JS reads the trust level
    // we just installed. Without this, `editor.workspaceTrustLevel()`
    // from the plugin returns whatever was current when the editor was
    // first constructed (the harness's permissive default), which would
    // make env-manager's `isTrusted()` check disagree with the gate's
    // actual decision.
    harness.editor_mut().update_plugin_state_snapshot();
    harness.editor_mut().fire_plugins_loaded_hook();
    harness.render().unwrap();
    harness
}

/// A bare `.venv` is executable content: activating it runs the repo's Python,
/// which auto-executes any `.pth`/`sitecustomize.py` shipped inside the venv (a
/// documented malware-drop vector), and a virtualenv is a module-namespace
/// boundary — NOT a security boundary. So a venv-only folder raises the single
/// core trust modal like any other executable content; it must NOT silently
/// auto-trust. Once trusted, the path-only env activates with no second prompt.
#[test]
fn test_venv_prompts_then_activates_on_trust() {
    let tmp = TempDir::new().unwrap();
    let project = tmp.path().to_path_buf();
    make_venv(&project);
    setup_env_manager(&project);

    let mut harness = boot_harness_like_main(120, 40, project);

    // venv-only raises the core trust modal naming the marker — not silently
    // trusted, not silently activated.
    harness
        .wait_until(|h| {
            let s = h.screen_to_string();
            s.contains("SECURITY WARNING") && s.contains("Detected: .venv")
        })
        .unwrap();
    let s = harness.screen_to_string();
    assert!(
        !s.contains("Trust & activate") && !s.contains("Environment detected"),
        "venv must not surface the env-manager popup"
    );

    // Trust the folder: mnemonic 't' selects "Trust", Enter confirms.
    harness
        .send_key(KeyCode::Char('t'), KeyModifiers::NONE)
        .unwrap();
    harness
        .send_key(KeyCode::Enter, KeyModifiers::NONE)
        .unwrap();

    // Now the path-only env activates silently (no second prompt) via the
    // `trust_changed` hook.
    harness
        .wait_until(|h| {
            let s = h.screen_to_string();
            !s.contains("SECURITY WARNING")
                && s.contains(".venv")
                && (s.contains("Activating") || s.contains("active"))
        })
        .unwrap();
}

/// A session opened through the Orchestrator — a window created via
/// `create_window_with_terminal`, the path the "New Session (Local)" flow and
/// the session dock drive — must run the same workspace-trust decision a direct
/// `fresh <dir>` launch gets. For a `.venv` that means: raise the core trust
/// modal (a venv is executable content, not silently trusted), and once trusted
/// activate the env in the new window via the `trust_changed` hook.
///
/// Regression (issue #2355): the orchestrator window used to bypass
/// `maybe_prompt_workspace_trust` entirely, so the session never got the trust
/// decision and never activated. This drives the create-window action and
/// asserts the modal appears in the new window and that trusting activates the
/// env there.
#[test]
fn test_orchestrator_session_prompts_then_activates_venv() {
    if !pty_available() {
        eprintln!("Skipping orchestrator env activation test: PTY not available");
        return;
    }

    let tmp = TempDir::new().unwrap();
    // Launch project: plain (no env marker) so booting it activates nothing —
    // env-manager loads here and serves every window in the editor.
    let launch = tmp.path().join("launch");
    fs::create_dir_all(&launch).unwrap();
    setup_env_manager(&launch);
    // The project opened *through the orchestrator*: a path-only `.venv`.
    let venv_proj = tmp.path().join("venvproj");
    fs::create_dir_all(&venv_proj).unwrap();
    make_venv(&venv_proj);

    let mut harness = boot_harness_like_main(120, 40, launch);

    // Orchestrator "New Session (Local)" for the venv project — the same call
    // the plugin dispatcher makes for `createWindowWithTerminal`. The new
    // window is born under its own per-session local authority.
    let born = harness.editor().local_session_authority(&venv_proj);
    harness
        .editor_mut()
        .create_window_with_terminal(
            venv_proj.clone(),
            "venvproj".into(),
            Some(venv_proj.clone()),
            // A harmless long-running child so the PTY doesn't exit immediately.
            Some(vec!["sh".into(), "-c".into(), "sleep 60".into()]),
            None,
            born,
            None,
        )
        .expect("create orchestrator session window");

    // The new session gets the trust decision just like a direct launch: the
    // core modal fires in the new window, naming the marker. (Before the
    // regression fix it never appeared — the session bypassed the prompt.)
    harness
        .wait_until(|h| {
            let s = h.screen_to_string();
            s.contains("SECURITY WARNING") && s.contains("Detected: .venv")
        })
        .unwrap();

    // Trust it (mnemonic 't' + Enter); the path-only env then activates in the
    // new window via `trust_changed`.
    harness
        .send_key(KeyCode::Char('t'), KeyModifiers::NONE)
        .unwrap();
    harness
        .send_key(KeyCode::Enter, KeyModifiers::NONE)
        .unwrap();
    harness
        .wait_until(|h| {
            let s = h.screen_to_string();
            !s.contains("SECURITY WARNING")
                && s.contains(".venv")
                && (s.contains("Activating") || s.contains("active"))
        })
        .unwrap();
}

/// `.envrc` raises the *core* trust modal — the single trust prompt, same as
/// any other executable content — with concrete framing naming the marker.
/// Picking "Trust" elevates trust, and env-manager activates direnv in
/// response to the resulting `trust_changed` hook. There is no separate,
/// plugin-owned "Trust & activate" popup (that duplicate was removed).
#[test]
fn test_envrc_raises_core_trust_modal_and_activates_on_trust() {
    let tmp = TempDir::new().unwrap();
    let project = tmp.path().to_path_buf();
    make_envrc(&project);
    setup_env_manager(&project);

    let mut harness = boot_harness_like_main(140, 40, project);

    // The core trust modal fires for `.envrc`, naming the marker concretely.
    harness
        .wait_until(|h| {
            let s = h.screen_to_string();
            s.contains("SECURITY WARNING") && s.contains("Detected: .envrc")
        })
        .unwrap();
    // The old plugin-owned combined popup must NOT appear — that's the
    // duplicate this change removed.
    let snapshot = harness.screen_to_string();
    assert!(
        !snapshot.contains("Environment detected") && !snapshot.contains("Trust & activate"),
        ".envrc must not surface the env-manager trust popup (duplicate removed)"
    );

    // Select "Trust this folder" (mnemonic `t`), then confirm with Enter.
    harness
        .send_key(KeyCode::Char('t'), KeyModifiers::NONE)
        .unwrap();
    harness
        .send_key(KeyCode::Enter, KeyModifiers::NONE)
        .unwrap();

    // Trusting fires `trust_changed`; env-manager activates direnv in
    // response. The activation message from `applyActivation` includes
    // "Activating direnv", and the modal is gone.
    harness
        .wait_until(|h| {
            let s = h.screen_to_string();
            !s.contains("SECURITY WARNING") && s.contains("Activating direnv")
        })
        .unwrap();
}

/// A project manifest (`Cargo.toml`) raises the core trust modal with
/// *concrete framing* — the body must name the actual marker so the user
/// knows why they're being asked.
#[test]
fn test_cargo_toml_raises_trust_modal_with_concrete_framing() {
    let tmp = TempDir::new().unwrap();
    let project = tmp.path().to_path_buf();
    make_cargo_toml(&project);
    setup_env_manager(&project);

    let mut harness = boot_harness_like_main(140, 40, project);

    harness
        .wait_until(|h| {
            let s = h.screen_to_string();
            s.contains("SECURITY WARNING") && s.contains("Detected: Cargo.toml")
        })
        .unwrap();
    // env popup must NOT be present — manifest-only flow is the core modal.
    let s = harness.screen_to_string();
    assert!(
        !s.contains("Environment detected"),
        "Cargo.toml-only folder must not fire the env popup"
    );
}

/// T19 — the user opens a folder that fires the trust modal, then quits
/// without picking a row (Ctrl+Q). The contract: trust must NOT be
/// recorded, so the next open re-prompts. Before the fix in
/// `0424342`, the initial-Restricted seed was being persisted, so this
/// test would observe the modal absent on second open — a clear
/// indicator that the cancel had been (incorrectly) interpreted as a
/// choice.
///
/// Critical setup detail: the two boots share a `DirectoryContext` so
/// any `trust.json` written by boot 1 is visible to boot 2. Without
/// this, each harness gets a fresh `data_dir` and the test would pass
/// even with the bug present.
#[test]
fn test_quit_cancels_trust_modal_without_recording_decision() {
    let tmp = TempDir::new().unwrap();
    let project = tmp.path().to_path_buf();
    make_cargo_toml(&project);
    setup_env_manager(&project);

    // Shared state dir so trust persistence (if any) survives between
    // the two harness lifetimes.
    let state_tmp = TempDir::new().unwrap();
    let dir_context = DirectoryContext::for_testing(state_tmp.path());

    {
        let mut harness =
            boot_with_dir_context(140, 40, project.clone(), Some(dir_context.clone()));
        harness
            .wait_until(|h| {
                let s = h.screen_to_string();
                s.contains("SECURITY WARNING") && s.contains("Detected: Cargo.toml")
            })
            .unwrap();
        // Quit via Ctrl+Q. The trust modal's non-cancellable variant
        // binds the global quit key to "close + quit editor."
        harness
            .send_key(KeyCode::Char('q'), KeyModifiers::CONTROL)
            .unwrap();
        // Harness drops here; equivalent to a clean exit. Any decision
        // the editor recorded is now on disk under `dir_context`.
    }

    // Boot a *second* harness with the same dir_context. The trust
    // store reads from the same `<data_dir>/workspaces/<encoded>/
    // trust.json`, so any persisted decision from boot 1 short-
    // circuits `maybe_prompt_workspace_trust`. If the modal appears
    // again, no decision was recorded — the fix is correct. If the
    // modal does NOT appear, the quit was (incorrectly) interpreted
    // as choosing Restricted.
    let mut harness2 = boot_with_dir_context(140, 40, project, Some(dir_context));
    harness2
        .wait_until(|h| {
            let s = h.screen_to_string();
            s.contains("SECURITY WARNING") && s.contains("Detected: Cargo.toml")
        })
        .unwrap();
}
